lib/SILOptimizer/SemanticARC/BorrowScopeOpts.cpp

```diff
@@ -31,8 +31,20 @@ bool SemanticARCOptVisitor::visitBeginBorrowInst(BeginBorrowInst *bbi) {
 
   // Lexical borrow scopes must remain in order to ensure that value lifetimes
   // are not observably shortened.
-  if (bbi->isLexical())
-    return false;
+  if (bbi->isLexical()) {
+    auto value = bbi->getOperand();
+    auto shouldBail = [&]() {
+      if (auto *outerBBI = dyn_cast<BeginBorrowInst>(value)) {
+        return !outerBBI->isLexical();
+      }
+      if (auto *arg = dyn_cast<SILFunctionArgument>(value)) {
+        return arg->getOwnershipKind() == OwnershipKind::Guaranteed;
```

Comment:
I think this should be `arg->getOwnershipKind() == OwnershipKind::Guaranteed`

Reply:
Fixed: extracted lambda to static function with clearer name.  

Aside: It seems like we may want to do the same check in the other places where we are bailing on lexical begin borrows (TempRValueOpts, etc).  At that point we can make that static function more broadly visible.